File: 프로그래머스/3/43163. 단어 변환/단어 변환.js
function solution(begin, target, words) {
    let answer = 51

    if (words.includes(target)) {
        searchWords(begin, [], 0);
    }

    function searchWords(curWord, visited, count) {
        if (curWord === target) {
            answer = Math.min(answer, count)
            return
        }

        for (let i = 0; i < words.length; i++) {
            if (visited[i] !== true && checkNextWord(curWord, words[i])) {
                visited[i] = true
                searchWords(words[i], visited, count+1)
                visited[i] = false
            }
        }
    }

    function checkNextWord(cur, next) {
        let differentWords = 0

        for (let i = 0; i < cur.length; i++) {
            if (cur[i] !== next[i]) {
                differentWords++
            }
        }
        return differentWords === 1 ? true : false
    }
    return answer === 51 ? 0 : answer
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
def solution(begin, target, words):
    answer = [51]

    def check_next_word(cur, nxt):
        diff_count = sum(1 for c, n in zip(cur, nxt) if c != n)
        return diff_count == 1

    def search_words(cur_word, visited, count):
        if cur_word == target:
            answer[0] = min(answer[0], count)
            return

        for i, word in enumerate(words):
            if not visited[i] and check_next_word(cur_word, word):
                visited[i] = True
                search_words(word, visited, count + 1)
                visited[i] = False

    if target in words:
        search_words(begin, [False] * len(words), 0)

    return 0 if answer[0] == 51 else answer[0]